Improve and develop automate_bindings.py

zer0main opened the pull request.

zer0main committed:
Automate bindings: use ' instead of " everywhere

zer0main committed:
Complement TYPE_TO_LUA_FUNCS and TYPE_FROM_LUA_FUNCS

zer0main committed:
Add new dicts for storing stuff related to problematic types

Problematic type is a type which isn't built-in and which
isn't descendant of WObject.

zer0main committed:
automate_bindings.py: add findCorrespondingKeyInDict()

zer0main committed:
automate_bindings.py: rewrite getBuiltinType()

Return corresponding built-in type in case of problematic input;
return corresponding built-in type in case of built-in input;
return '' (False) otherwise.

zer0main committed:
getBuiltinTypeArgument(): add support of problematic types

zer0main committed:
returnValues(): add support of problematic types

starius commented on tools/automate_bindings.py:
```diff
@@ -70,11 +82,29 @@ def getSelf(module_name):
 
+def findCorrespondingKeyInDict(dictionary, full_key):
+    for key in dictionary:
+        if full_key.find(key) is not -1:
```
1. Numbers should be compared with `==`, not with `is`.
2. `if key in full_key:`

zer0main replied:
Done :+1:

starius commented on tools/automate_bindings.py:
```diff
@@ -111,17 +141,38 @@ def returnValue(return_type):
             func_name = 'luawt_toLua'
-        return RETURN_CALLS_TEMPLATE % func_name
+        # Method to convert problematic type to built-in.
+        # Empty by defualt.
```
typo

zer0main replied:
Fixed :+1:

starius commented on tools/automate_bindings.py:
```diff
@@ -111,17 +141,38 @@ def returnValue(return_type):
+        if problematic_type:
+            convert_f += '.'
+            convert_f += PROBLEMATIC_TYPES_CONVERSIONS[problematic_type]
+            convert_f += '()'
```
Can you do it in one line and use `=` instead of `+=`?

zer0main replied:
Ok, done.

starius commented on tools/automate_bindings.py:
```diff
@@ -271,10 +322,10 @@ def writeSourceToFile(module_name, source):
+        writeToFile(parameter['filename'], addItem(
+            parameter['pattern'],
+            parameter['module_str'],
             content
```
`,`

zer0main replied:
Done :+1:

zer0main committed:
Fix typo

zer0main committed:
automate_bindings.py: improve style

zer0main committed:
returnValue(): initialise convert_f in one line

Use `=` instead of `+=`.

zer0main committed:
automate_bindings.py: do not use `is` when comparing numbers

zer0main committed:
Add new option --module-only

If it's specified, globals.hpp, init.cpp and rockspec aren't changed.

zer0main committed:
Filter all non-public methods

zer0main committed:
getMethodsAndBases(): find main class properly

zer0main committed:
getConstructor(): find main class properly

zer0main committed:
generateModuleFunc(): use base name, not a type

zer0main committed:
Merge two dicts with the same keys

TYPE_FROM_LUA_FUNCS and TYPE_TO_LUA_FUNCS -->
BUILTIN_TYPES_CONVERTERS (key --> tuple(val1, val2)).

zer0main committed:
Add isConstReference() function

zer0main committed:
Add checkReturnType() function

zer0main committed:
generateConstructor(): rename some vars to clarify the func

zer0main committed:
implementLuaCFunction(): do not bind unsupported funcs

zer0main committed:
returnValue(): support references

zer0main committed:
Add clearType() func

zer0main committed:
Add getArgsStr() func

zer0main committed:
Use getArgsStr() in callWtFunction() and callWtConstructor()

zer0main committed:
getComplexArgument(): pass type_t instead of str

Clear options['argument_type'], in order to pass it as template
parameter in C++ code.

zer0main committed:
Do not call checkReturnType() from implementLuaCFunction()

All the checks will be carried out in getMethodsAndBases() and
getConstructor() functions.
Unsupported methods won't be added to the lists.

zer0main committed:
Add isTemplate() function

zer0main committed:
Add isWObjectsDescendant() function

We need to parse all Wt in order to check all the classes, so
if function can't find specified class, it prints warning and
returns True.

zer0main committed:
Add checkArgumentType() function

zer0main committed:
Move isConstReference() function at the top

It'll be widely used by checks functions.

zer0main committed:
checkReturnType(): filter methods with template arguments

Also move checkReturnType() at the top.

zer0main committed:
Add addEnum() function

Consider C++ `enum` as a built-in type:
 - push it to Lua as integer
 - get it from Lua using lua_tointeger with applied `static_cast`

zer0main committed:
getBuiltinTypeArgument(): add some options for enums

zer0main committed:
Add checkWtFunction()

Unite all the checks.

zer0main committed:
getMethodsAndBases(): filter methods using checkWtFunction()

zer0main committed:
getConstructor(): use checkWtFunction() to check constructor

zer0main committed:
Travis: now it should compile

Remove TODO for zer0main, because it was successfully completed.

zer0main committed:
Add getBuiltinTypeFromProblematic()

zer0main committed:
Develop problematic types' support

Dicts: problematic_type ---> (convert_function(), its arg/return_type);
returnValue(): use getBuiltinTypeFromProblematic();
getBuiltinType(): return closest built-in type in case of problematic
input, search for it using cycle.

zer0main committed:
Add getNumberOfTransitionalTypes()

Count number of types between the given problematic and its
closest built-in type.

zer0main committed:
Add getProblematicFromBuiltin()

zer0main committed:
getBuiltinTypeArgument(): use getProblematicFromBuiltin()

zer0main committed:
Add Wt::WString to the list of problematic types

zer0main committed:
Add function loadAdditionalChunk()

Usage: when unable to find class, try to load it's module directly
by using this function.

zer0main committed:
getConstructor(): find suitable constructor

zer0main committed:
Add function getClassStr()

zer0main committed:
Add function getClass()

zer0main committed:
Add function isDescendantLogic()

zer0main committed:
Add function isDescendantOf()

zer0main committed:
Remove isWObjectsDescendant(); use isDescendantOf() instead

zer0main committed:
getMethodsAndBases() --> getMethodsAndBase()

Find single correct base.
close #53

zer0main committed:
Add getClassNameFromModuleStr()

zer0main committed:
Take into account that base must be before child in init.cpp

Remove TODO for zer0main, because it was successfully completed.

zer0main committed:
Call generateModule() before addModuleToLists()

In order to check everything (and potentially raise Exceptions)
before writing anything to files.
Prevent generating meaningless pieces of code.

zer0main committed:
Global dicts: const char* --> char const *

pygccxml doesn't represent this type in this way.

zer0main committed:
Add function getArgType()

zer0main committed:
Use getArgType() in checkWtFunction() and implementLuaCFunction()

zer0main committed:
Do not bind abstract classes

zer0main committed:
Add isModule() func

zer0main committed:
Add addIncludeDir() func

zer0main committed:
Add getModulesFromFuncSig() func

zer0main committed:
Add getIncludes() func

zer0main committed:
generateIncludes(): do not forget to add some additional Wt includes

Use getIncludes() in generateModule(); pass it's result to
generateIncludes().

zer0main committed:
isDescendantOf(): only name of base is actually needed

It's very powerful optimization because we do not need to load
additional modules to get information about bases.

zer0main committed:
addItem(), init.cpp: iterate from the last class to the first

And check isDescendantOf(module_name, curr_class) instead of
isDescendantOf(curr_class, module_name) while iterating from
the first to the last.
We do need to load all the modules from init.cpp now.

zer0main committed:
checkWtFunction(): check that access_type is public

zer0main committed:
getConstructor(): filter non-public constructors

Use custom_matcher instead of cycle also.

zer0main committed:
parse(): cache xml files in order to speed up the process

zer0main committed:
Add function getAllModules()

zer0main committed:
bind(): add possibility of binding all the library

zer0main committed:
If --bind is not specified, scripts binds all the library

zer0main committed:
Bind only widgets

parseId() from globals.hpp won't work with classes which are not
descendants of WWidget.

zer0main committed:
Check that class is descendant of WWidget before checking all its methods

Small optimization.
